config/deprecation.go

```diff
@@ -0,0 +1,285 @@
+	deprecatedOptions := make([]string, 0)
+	for _, option := range info.Options {
+		switch option.Level {
+		case Warn:
+			prefix := "W! " + color.YellowString("DeprecationWarning")
+			printOptionDeprecationNotice(prefix, info.Name, option.Name, option.Since, option.Notice)
+		case Error:
+			prefix := "E! " + color.RedString("DeprecationError")
+			printOptionDeprecationNotice(prefix, info.Name, option.Name, option.Since, option.Notice)
+			deprecatedOptions = append(deprecatedOptions, option.Name)
+		}
+	}
+
+	if len(deprecatedOptions) > 0 {
+		return fmt.Errorf("plugin options %q deprecated", strings.Join(deprecatedOptions, ","))
+	}
+
+	return nil
+}
+
+func (c *Config) CollectDeprecationInfos() map[string][]PluginDeprecationInfo {
+	infos := make(map[string][]PluginDeprecationInfo)
+
+	infos["inputs"] = make([]PluginDeprecationInfo, 0)
+	for name, creator := range inputs.Inputs {
+		plugin := creator()
+		info := c.collectDeprecationInfo(name, plugin, true)
+
+		if info.Level != None || len(info.Options) > 0 {
+			infos["inputs"] = append(infos["inputs"], info)
+		}
+	}
+
+	infos["outputs"] = make([]PluginDeprecationInfo, 0)
+	for name, creator := range outputs.Outputs {
+		plugin := creator()
+		info := c.collectDeprecationInfo(name, plugin, true)
+
+		if info.Level != None || len(info.Options) > 0 {
+			infos["outputs"] = append(infos["outputs"], info)
+		}
+	}
+
+	infos["processors"] = make([]PluginDeprecationInfo, 0)
+	for name, creator := range processors.Processors {
+		plugin := creator()
+		info := c.collectDeprecationInfo(name, plugin, true)
+
+		if info.Level != None || len(info.Options) > 0 {
+			infos["processors"] = append(infos["processors"], info)
+		}
+	}
+
+	infos["aggregators"] = make([]PluginDeprecationInfo, 0)
+	for name, creator := range aggregators.Aggregators {
+		plugin := creator()
+		info := c.collectDeprecationInfo(name, plugin, true)
+
+		if info.Level != None || len(info.Options) > 0 {
+			infos["aggregators"] = append(infos["aggregators"], info)
+		}
+	}
+
+	return infos
+}
+
+func (c *Config) PrintDeprecationList(infos []PluginDeprecationInfo) {
+	sort.Slice(infos, func(i, j int) bool { return infos[i].Name < infos[j].Name })
+
+	for _, info := range infos {
+		switch info.Level {
+		case Warn, Error:
+			_, _ = fmt.Printf("  %-40s %-5s since %5s %s\n", info.Name, info.Level, info.Since, info.Notice)
+		}
+
+		if len(info.Options) < 1 {
+			continue
+		}
+		sort.Slice(info.Options, func(i, j int) bool { return info.Options[i].Name < info.Options[j].Name })
+		for _, option := range info.Options {
+			_, _ = fmt.Printf("  %-40s %-5s since %5s %s\n", info.Name+"/"+option.Name, option.Level, option.Since, option.Notice)
+		}
+	}
+}
+
+func (c *Config) getDeprecationEscalation(since string) Escalation {
+	sinceMajor, sinceMinor := ParseVersion(since)
+	if c.VersionMajor > sinceMajor {
+		return Error
+	}
+	if c.VersionMajor == sinceMajor && c.VersionMinor >= sinceMinor {
+		return Warn
+	}
```

Comment:
If we deprecate a plugin in say 1.21 and then move to 2.0 right after, then by this logic, that plugin is no longer supported and we have not given the user any time to transition off. I would rather us say something like "in X version this plugin will no longer be supported."

Reply:
See my comment above.